File: android/app/src/main/java/com/varun/ignyt/healthconnect/HealthConnectPlugin.kt
package com.varun.ignyt.healthconnect

import androidx.health.connect.client.HealthConnectClient
import androidx.health.connect.client.records.ExerciseSessionRecord
import com.getcapacitor.JSObject
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONArray
import java.time.Instant

/**
 * Capacitor bridge for Health Connect, scoped to IGNYT's current requirements: read
 * steps/heart-rate/weight/active-calories/distance, write exercise sessions and weight.
 * Every method resolves {"success": true, "data": ...} or {"success": false, "error": "..."}
 * -- never rejects -- so the JS side always gets a normal resolved promise.
 *
 * Permissions are requested ONLY from requestPermissions(), never automatically. Every
 * read/write method checks permissions first and returns a clear error (not an automatic
 * permission prompt) if they're missing -- matches "request only after explicit user action".
 */
@CapacitorPlugin(name = "HealthConnect")
class HealthConnectPlugin : com.getcapacitor.Plugin() {

    private lateinit var manager: HealthConnectManager
    private val pluginScope = CoroutineScope(Dispatchers.Main)

    override fun load() {
        manager = HealthConnectManager(context)
    }

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        try {
            val status = manager.sdkStatus()
            // Health Connect 1.1.0 does not expose SDK_AVAILABLE_UPDATE_REQUIRED as a usable
            // constant here (that three-way distinction is a later-version addition) -- rather
            // than reference something that doesn't compile against 1.1.0, this collapses to
            // a plain available/unavailable check. Still fully correct and still drives the
            // "open Play Store when unavailable" flow; it just can't distinguish "not
            // installed" from "installed but needs updating" until upgrading past 1.1.0.
            val data = JSObject().apply {
                put("available", status == HealthConnectClient.SDK_AVAILABLE)
                put("status", if (status == HealthConnectClient.SDK_AVAILABLE) "AVAILABLE" else "UNAVAILABLE")
            }
            resolveSuccess(call, data)
        } catch (e: Exception) {
            resolveError(call, "isAvailable failed: ${e.message}")
        }
    }

    /** Separate from isAvailable() on purpose: opening the Play Store is a real user-facing
     *  action, so it only happens when explicitly called, not as a side effect of a status check. */
    @PluginMethod
    fun openHealthConnectInstall(call: PluginCall) {
        try {
            activity.startActivity(manager.playStoreInstallIntent())
            resolveSuccess(call, JSObject().apply { put("opened", true) })
        } catch (e: Exception) {
            resolveError(call, "Could not open Play Store: ${e.message}")
        }
    }

    // Capacitor's own Plugin base class already declares requestPermissions(PluginCall) as an
    // overridable hook -- it's the documented extension point for plugins whose permission
    // model isn't standard Android runtime permissions (like Health Connect's). Overriding it
    // is the correct fix, not a naming collision to work around: it keeps the JS-facing method
    // name exactly "requestPermissions" with no remapping needed.
    @PluginMethod
    override fun requestPermissions(call: PluginCall) {
        if (!manager.isAvailable()) {
            resolveError(call, "Health Connect is not available on this device.")
            return
        }
        pluginScope.launch {
            try {
                if (manager.hasAllPermissions()) {
                    resolveSuccess(call, JSObject().apply { put("granted", true) })
                    return@launch
                }
                val contract = manager.permissionRequestContract()
                val intent = contract.createIntent(context, manager.allPermissions)
                saveCall(call)
                startActivityForResult(call, intent, "permissionCallback")
            } catch (e: Exception) {
                resolveError(call, "requestPermissions failed: ${e.message}")
            }
        }
    }

    @ActivityCallback
    private fun permissionCallback(call: PluginCall?, result: androidx.activity.result.ActivityResult) {
        if (call == null) return
        pluginScope.launch {
            try {
                val granted = manager.grantedPermissions()
                val data = JSObject().apply {
                    put("granted", granted.containsAll(manager.allPermissions))
                    put("grantedPermissions", JSONArray(granted.toList()))
                }
                resolveSuccess(call, data)
            } catch (e: Exception) {
                resolveError(call, "Permission result handling failed: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun getPermissionStatus(call: PluginCall) {
        if (!manager.isAvailable()) {
            resolveSuccess(call, JSObject().apply { put("granted", false); put("available", false) })
            return
        }
        pluginScope.launch {
            try {
                val granted = manager.grantedPermissions()
                resolveSuccess(call, JSObject().apply {
                    put("available", true)
                    put("granted", granted.containsAll(manager.allPermissions))
                    put("grantedPermissions", JSONArray(granted.toList()))
                })
            } catch (e: Exception) {
                resolveError(call, "getPermissionStatus failed: ${e.message}")
            }
        }
    }

    @PluginMethod
    fun revokePermissions(call: PluginCall) {
        pluginScope.launch {
            try {
                manager.revokeAllPermissions()
                resolveSuccess(call, JSObject().apply { put("revoked", true) })
            } catch (e: Exception) {
                resolveError(call, "revokePermissions failed: ${e.message}")
            }
        }
    }

    private suspend fun ensurePermissions(call: PluginCall): Boolean {
        if (!manager.isAvailable()) {
            resolveError(call, "Health Connect is not available on this device.")
            return false
        }
        if (manager.hasAllPermissions()) return true
        resolveError(call, "Health Connect permissions have not been granted yet. Call requestPermissions() first.")
        return false
    }

    @PluginMethod
    fun getTodaySteps(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getTodaySteps() } } }

    @PluginMethod
    fun getHeartRate(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getHeartRate() } } }

    @PluginMethod
    fun getLatestWeight(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestWeight() } } }

    @PluginMethod
    fun getTodayActiveCalories(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getTodayActiveCalories() } } }

    @PluginMethod
    fun getTodayDistance(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getTodayDistance() } } }

    // ADDED
    @PluginMethod
    fun getTodayWorkoutCount(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getTodayExerciseSessionCount() } } }

    // ADDED -- the 10 newly requested metrics
    @PluginMethod
    fun getRespiratoryRate(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestRespiratoryRate() } } }

    @PluginMethod
    fun getOxygenSaturation(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestOxygenSaturation() } } }

    @PluginMethod
    fun getBloodPressure(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestBloodPressure() } } }

    @PluginMethod
    fun getBodyTemperature(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestBodyTemperature() } } }

    @PluginMethod
    fun getBodyFat(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestBodyFat() } } }

    @PluginMethod
    fun getHeight(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestHeight() } } }

    @PluginMethod
    fun getLeanBodyMass(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestLeanBodyMass() } } }

    @PluginMethod
    fun getBasalMetabolicRate(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestBasalMetabolicRate() } } }

    @PluginMethod
    fun getTodayHydration(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getTodayHydration() } } }

    @PluginMethod
    fun getTodayNutrition(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getTodayNutrition() } } }

    // ADDED: trend data + sleep, for the redesigned Home dashboard's mini charts
    @PluginMethod
    fun getSleepSummary(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolve(call) { manager.getLatestSleepSession() } } }

    @PluginMethod
    fun getHeartRateHistory(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolveArray(call) { manager.getHeartRateRecentSeries() } } }

    @PluginMethod
    fun getStepsHistory(call: PluginCall) { pluginScope.launch { if (ensurePermissions(call)) safeResolveArray(call) { manager.getStepsHistory7Days() } } }

    @PluginMethod
    fun getWeightHistory(call: PluginCall) {
        val days = call.getInt("days", 90) ?: 90
        pluginScope.launch { if (ensurePermissions(call)) safeResolveArray(call) { manager.getWeightHistory(days) } }
    }

    private suspend fun safeResolveArray(call: PluginCall, block: suspend () -> JSONArray) {
        try {
            resolveSuccess(call, JSObject().apply { put("items", block()) })
        } catch (e: Exception) {
            resolveError(call, "Read failed: ${e.message}")
        }
    }

    private suspend fun safeResolve(call: PluginCall, block: suspend () -> org.json.JSONObject) {
        try {
            resolveSuccess(call, JSObject.fromJSONObject(block()))
        } catch (e: Exception) {
            resolveError(call, "Read failed: ${e.message}")
        }
    }

    /** One call for the dashboard/Sync button: fetches all 5 "today" reads. Individual
     *  failures don't abort the whole thing -- each field is null if its own read failed. */
    @PluginMethod
    fun syncNow(call: PluginCall) {
        pluginScope.launch {
            if (!ensurePermissions(call)) return@launch
            val data = JSObject()
            // Every field is independently wrapped in safeOrNull -- one metric failing
            // (missing permission for a newly-added type, a transient read error, etc.)
            // never prevents the others from returning. No single bad read blanks the screen.
            data.put("steps", safeOrNull { manager.getTodaySteps() }?.let { JSObject.fromJSONObject(it) })
            data.put("heartRate", safeOrNull { manager.getHeartRate() }?.let { JSObject.fromJSONObject(it) })
            data.put("weight", safeOrNull { manager.getLatestWeight() }?.let { JSObject.fromJSONObject(it) })
            data.put("activeCalories", safeOrNull { manager.getTodayActiveCalories() }?.let { JSObject.fromJSONObject(it) })
            data.put("distance", safeOrNull { manager.getTodayDistance() }?.let { JSObject.fromJSONObject(it) })
            data.put("workouts", safeOrNull { manager.getTodayExerciseSessionCount() }?.let { JSObject.fromJSONObject(it) })
            data.put("sleep", safeOrNull { manager.getLatestSleepSession() }?.let { JSObject.fromJSONObject(it) }) // ADDED
            data.put("steps7Days", safeOrNull { manager.getStepsHistory7Days() }) // ADDED
            data.put("heartRateSeries", safeOrNull { manager.getHeartRateRecentSeries() }) // ADDED
            data.put("weightHistory", safeOrNull { manager.getWeightHistory(90) }) // ADDED
            // ADDED -- the 10 newly requested metrics, each independently wrapped so one
            // missing permission or empty result never affects any other field
            data.put("respiratoryRate", safeOrNull { manager.getLatestRespiratoryRate() }?.let { JSObject.fromJSONObject(it) })
            data.put("oxygenSaturation", safeOrNull { manager.getLatestOxygenSaturation() }?.let { JSObject.fromJSONObject(it) })
            data.put("bloodPressure", safeOrNull { manager.getLatestBloodPressure() }?.let { JSObject.fromJSONObject(it) })
            data.put("bodyTemperature", safeOrNull { manager.getLatestBodyTemperature() }?.let { JSObject.fromJSONObject(it) })
            data.put("bodyFat", safeOrNull { manager.getLatestBodyFat() }?.let { JSObject.fromJSONObject(it) })
            data.put("height", safeOrNull { manager.getLatestHeight() }?.let { JSObject.fromJSONObject(it) })
            data.put("leanBodyMass", safeOrNull { manager.getLatestLeanBodyMass() }?.let { JSObject.fromJSONObject(it) })
            data.put("basalMetabolicRate", safeOrNull { manager.getLatestBasalMetabolicRate() }?.let { JSObject.fromJSONObject(it) })
            data.put("hydration", safeOrNull { manager.getTodayHydration() }?.let { JSObject.fromJSONObject(it) })
            data.put("nutrition", safeOrNull { manager.getTodayNutrition() }?.let { JSObject.fromJSONObject(it) })
            data.put("syncedAt", System.currentTimeMillis())
            resolveSuccess(call, data)
        }
    }

    private suspend fun <T> safeOrNull(block: suspend () -> T): T? = try { block() } catch (e: Exception) { null }

    /** ignytWorkoutId is required, not optional -- it's what makes duplicate-export
     *  prevention actually work (see HealthConnectManager's clientRecordId comment). */
    @PluginMethod
    fun saveWorkout(call: PluginCall) {
        pluginScope.launch {
            if (!ensurePermissions(call)) return@launch
            try {
                val startMillis = call.getString("startTime")?.toLongOrNull()
                val endMillis = call.getString("endTime")?.toLongOrNull()
                val title = call.getString("title") ?: "Workout"
                val ignytWorkoutId = call.getString("ignytWorkoutId")
                if (startMillis == null || endMillis == null || ignytWorkoutId.isNullOrBlank()) {
                    resolveError(call, "saveWorkout requires startTime, endTime, and ignytWorkoutId.")
                    return@launch
                }
                val exerciseType = mapIgnytExerciseType(call.getString("type"))
                val id = manager.saveWorkout(
                    Instant.ofEpochMilli(startMillis), Instant.ofEpochMilli(endMillis), title, exerciseType, ignytWorkoutId
                )
                resolveSuccess(call, JSObject().apply { put("recordId", id) })
            } catch (e: Exception) {
                resolveError(call, "saveWorkout failed: ${e.message}")
            }
        }
    }

    /** ignytBodyLogId required for the same reason. */
    @PluginMethod
    fun saveWeight(call: PluginCall) {
        pluginScope.launch {
            if (!ensurePermissions(call)) return@launch
            try {
                val weightKg = call.getDouble("weightKg")
                val ignytBodyLogId = call.getString("ignytBodyLogId")
                if (weightKg == null || ignytBodyLogId.isNullOrBlank()) {
                    resolveError(call, "saveWeight requires weightKg and ignytBodyLogId.")
                    return@launch
                }
                val timeMillis = call.getString("time")?.toLongOrNull() ?: System.currentTimeMillis()
                val id = manager.saveWeight(weightKg, Instant.ofEpochMilli(timeMillis), ignytBodyLogId)
                resolveSuccess(call, JSObject().apply { put("recordId", id) })
            } catch (e: Exception) {
                resolveError(call, "saveWeight failed: ${e.message}")
            }
        }
    }

    private fun resolveSuccess(call: PluginCall, data: JSObject) {
        call.resolve(JSObject().apply { put("success", true); put("data", data) })
    }

    private fun resolveError(call: PluginCall, message: String) {
        call.resolve(JSObject().apply { put("success", false); put("error", message) })
    }

    private fun mapIgnytExerciseType(ignytType: String?): Int = when (ignytType) {
        "race" -> ExerciseSessionRecord.EXERCISE_TYPE_HIGH_INTENSITY_INTERVAL_TRAINING
        "cardio" -> ExerciseSessionRecord.EXERCISE_TYPE_RUNNING
        else -> ExerciseSessionRecord.EXERCISE_TYPE_STRENGTH_TRAINING
    }
}

File: ios/App/App/HealthConnectPlugin.swift
import Foundation
import Capacitor
import HealthKit
import UIKit

/**
 The iOS half of IGNYT's health bridge.

 Registered as "HealthConnect" on purpose. The name is Android's, but renaming it would mean
 touching every call site in the web layer for no behavioural gain — the JS asks for health
 data, and this answers with HealthKit. The 29 methods, their arguments and their JSON shapes
 match android/.../HealthConnectPlugin.kt exactly, so www/health-connect.js runs unchanged.

 Like the Android plugin, no method ever rejects. Everything resolves either
 {"success": true, "data": ...} or {"success": false, "error": "..."} so the JS side always
 gets a normal resolved promise.

 Permissions are requested only from requestPermissions(), never as a side effect of a read.

 THREE METHODS CANNOT MEAN ON iOS WHAT THEY MEAN ON ANDROID. They are implemented honestly
 rather than faked; each says so in its own comment:
   getPermissionStatus   — iOS never reveals whether a READ was granted
   revokePermissions     — only the user can revoke, in Settings
   openHealthConnectInstall — nothing to install; Health is part of iOS
 */
@objc(HealthConnectPlugin)
public class HealthConnectPlugin: CAPPlugin, CAPBridgedPlugin {

    public let identifier = "HealthConnectPlugin"
    public let jsName = "HealthConnect"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openHealthConnectInstall", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPermissionStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "revokePermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getTodaySteps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getHeartRate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getLatestWeight", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getTodayActiveCalories", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getTodayDistance", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getTodayWorkoutCount", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getRespiratoryRate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getOxygenSaturation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBloodPressure", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBodyTemperature", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBodyFat", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getHeight", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getLeanBodyMass", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getBasalMetabolicRate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getTodayHydration", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getTodayNutrition", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getSleepSummary", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getHeartRateHistory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStepsHistory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getWeightHistory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "syncNow", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getInsights", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "saveWorkout", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "saveWeight", returnType: CAPPluginReturnPromise)
    ]

    private let manager = HealthKitManager()

    // MARK: - Envelope

    private func ok(_ call: CAPPluginCall, _ data: [String: Any]) {
        call.resolve(["success": true, "data": data])
    }

    private func fail(_ call: CAPPluginCall, _ message: String) {
        call.resolve(["success": false, "error": message])
    }

    /// Runs a read and resolves it in the standard envelope. The Android analogue of
    /// ensurePermissions() cannot exist here — iOS will not say whether reads are permitted —
    /// so what is checked instead is that the user has been through the sheet at least once.
    private func read(_ call: CAPPluginCall, _ work: @escaping () async throws -> [String: Any]) {
        guard manager.isAvailable else {
            fail(call, "HealthKit is not available on this device.")
            return
        }
        guard manager.hasRequestedAuthorization else {
            fail(call, "Health permissions have not been requested yet. Call requestPermissions() first.")
            return
        }
        Task {
            do { ok(call, try await work()) }
            catch { fail(call, "Read failed: \(error.localizedDescription)") }
        }
    }

    private func readList(_ call: CAPPluginCall, _ work: @escaping () async throws -> [[String: Any]]) {
        read(call) { ["items": try await work()] }
    }

    // MARK: - Availability and permissions

    @objc func isAvailable(_ call: CAPPluginCall) {
        let available = manager.isAvailable
        ok(call, ["available": available, "status": available ? "AVAILABLE" : "UNAVAILABLE"])
    }

    /// There is nothing to install — Health ships with iOS. Rather than silently succeeding,
    /// this opens the Health app, which is the nearest useful equivalent of the Android flow
    /// that sends the user to the Play Store listing.
    @objc func openHealthConnectInstall(_ call: CAPPluginCall) {
        guard let url = URL(string: "x-apple-health://") else {
            fail(call, "Could not open the Health app.")
            return
        }
        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:]) { opened in
                self.ok(call, ["opened": opened])
            }
        }
    }

    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        guard manager.isAvailable else {
            fail(call, "HealthKit is not available on this device.")
            return
        }
        Task {
            do {
                try await manager.requestAuthorization()
                // iOS reports nothing about reads, so `granted` here means the sheet completed
                // without error — not that any particular type was allowed.
                ok(call, ["granted": true, "partial": partialWrites(),
                          "grantedPermissions": grantedList(),
                          "readAuthorizationIsUnknowable": true])
            } catch {
                fail(call, "requestPermissions failed: \(error.localizedDescription)")
            }
        }
    }

    /**
     iOS will not tell an app whether a READ permission was granted. `authorizationStatus(for:)`
     answers only for types the app can write, and a denied read is indistinguishable from an
     empty data set — that is a deliberate privacy design, not a gap to work around.

     So `granted` reports whether the authorisation sheet has been completed, which is the only
     fact available, and `partial` is computed from the two write types where the answer is
     real. `readAuthorizationIsUnknowable` is added so the JS can tell the difference between
     this and Android's answer rather than assuming they mean the same thing.
     */
    @objc func getPermissionStatus(_ call: CAPPluginCall) {
        guard manager.isAvailable else {
            ok(call, ["granted": false, "available": false])
            return
        }
        ok(call, [
            "available": true,
            "granted": manager.hasRequestedAuthorization,
            "partial": partialWrites(),
            "grantedPermissions": grantedList(),
            "readAuthorizationIsUnknowable": true
        ])
    }

    /// No API exists for this on iOS; permissions are revoked by the user in
    /// Settings › Privacy & Security › Health › IGNYT. Reporting failure is the truthful
    /// answer — resolving success would tell the UI that data access had stopped when it had not.
    @objc func revokePermissions(_ call: CAPPluginCall) {
        fail(call, "iOS does not allow an app to revoke Health permissions. Open Settings › Privacy & Security › Health › IGNYT to change them.")
    }

    private func grantedList() -> [String] {
        manager.writeAuthorizationStatuses().filter { $0.value == "granted" }.map { $0.key }
    }

    private func partialWrites() -> Bool {
        manager.writeAuthorizationStatuses().values.contains { $0 != "granted" }
    }

    // MARK: - Reads

    @objc func getTodaySteps(_ call: CAPPluginCall)          { read(call) { try await self.manager.todaySteps() } }
    @objc func getHeartRate(_ call: CAPPluginCall)           { read(call) { try await self.manager.heartRate() } }
    @objc func getLatestWeight(_ call: CAPPluginCall)        { read(call) { try await self.manager.latestWeight() } }
    @objc func getTodayActiveCalories(_ call: CAPPluginCall) { read(call) { try await self.manager.todayActiveCalories() } }
    @objc func getTodayDistance(_ call: CAPPluginCall)       { read(call) { try await self.manager.todayDistance() } }
    @objc func getTodayWorkoutCount(_ call: CAPPluginCall)   { read(call) { try await self.manager.todayWorkoutCount() } }
    @objc func getRespiratoryRate(_ call: CAPPluginCall)     { read(call) { try await self.manager.latestRespiratoryRate() } }
    @objc func getOxygenSaturation(_ call: CAPPluginCall)    { read(call) { try await self.manager.latestOxygenSaturation() } }
    @objc func getBloodPressure(_ call: CAPPluginCall)       { read(call) { try await self.manager.latestBloodPressure() } }
    @objc func getBodyTemperature(_ call: CAPPluginCall)     { read(call) { try await self.manager.latestBodyTemperature() } }
    @objc func getBodyFat(_ call: CAPPluginCall)             { read(call) { try await self.manager.latestBodyFat() } }
    @objc func getHeight(_ call: CAPPluginCall)              { read(call) { try await self.manager.latestHeight() } }
    @objc func getLeanBodyMass(_ call: CAPPluginCall)        { read(call) { try await self.manager.latestLeanBodyMass() } }
    @objc func getBasalMetabolicRate(_ call: CAPPluginCall)  { read(call) { try await self.manager.latestBasalMetabolicRate() } }
    @objc func getTodayHydration(_ call: CAPPluginCall)      { read(call) { try await self.manager.todayHydration() } }
    @objc func getTodayNutrition(_ call: CAPPluginCall)      { read(call) { try await self.manager.todayNutrition() } }
    @objc func getSleepSummary(_ call: CAPPluginCall)        { read(call) { try await self.manager.latestSleepSession() } }

    @objc func getHeartRateHistory(_ call: CAPPluginCall)    { readList(call) { try await self.manager.heartRateSeries() } }
    @objc func getStepsHistory(_ call: CAPPluginCall)        { readList(call) { try await self.manager.stepsHistory() } }

    @objc func getWeightHistory(_ call: CAPPluginCall) {
        let days = call.getInt("days") ?? 90
        readList(call) { try await self.manager.weightHistory(days: days) }
    }

    // MARK: - Sync

    /// One call for the dashboard's Sync button. Every field is fetched independently, so one
    /// metric failing — a type the user declined, a transient error — leaves the rest intact
    /// rather than blanking the screen.
    @objc func syncNow(_ call: CAPPluginCall) {
        guard manager.isAvailable else {
            fail(call, "HealthKit is not available on this device.")
            return
        }
        Task {
            var data: [String: Any] = [
                "partialPermissions": self.partialWrites(),
                "grantedPermissions": self.grantedList(),
                "readAuthorizationIsUnknowable": true
            ]
            data["steps"]              = await self.orNull { try await self.manager.todaySteps() }
            data["heartRate"]          = await self.orNull { try await self.manager.heartRate() }
            data["weight"]             = await self.orNull { try await self.manager.latestWeight() }
            data["activeCalories"]     = await self.orNull { try await self.manager.todayActiveCalories() }
            data["distance"]           = await self.orNull { try await self.manager.todayDistance() }
            data["workouts"]           = await self.orNull { try await self.manager.todayWorkoutCount() }
            data["sleep"]              = await self.orNull { try await self.manager.latestSleepSession() }
            data["steps7Days"]         = await self.orNullList { try await self.manager.stepsHistory() }
            data["heartRateSeries"]    = await self.orNullList { try await self.manager.heartRateSeries() }
            data["weightHistory"]      = await self.orNullList { try await self.manager.weightHistory(days: 90) }
            data["respiratoryRate"]    = await self.orNull { try await self.manager.latestRespiratoryRate() }
            data["oxygenSaturation"]   = await self.orNull { try await self.manager.latestOxygenSaturation() }
            data["bloodPressure"]      = await self.orNull { try await self.manager.latestBloodPressure() }
            data["bodyTemperature"]    = await self.orNull { try await self.manager.latestBodyTemperature() }
            data["bodyFat"]            = await self.orNull { try await self.manager.latestBodyFat() }
            data["height"]             = await self.orNull { try await self.manager.latestHeight() }
            data["leanBodyMass"]       = await self.orNull { try await self.manager.latestLeanBodyMass() }
            data["basalMetabolicRate"] = await self.orNull { try await self.manager.latestBasalMetabolicRate() }
            data["hydration"]          = await self.orNull { try await self.manager.todayHydration() }
            data["nutrition"]          = await self.orNull { try await self.manager.todayNutrition() }
            data["syncedAt"] = Int(Date().timeIntervalSince1970 * 1000)
            self.ok(call, data)
        }
    }

    private func orNull(_ work: () async throws -> [String: Any]) async -> Any {
        (try? await work()) ?? NSNull()
    }

    private func orNullList(_ work: () async throws -> [[String: Any]]) async -> Any {
        (try? await work()) ?? NSNull()
    }

    // MARK: - Insights

    @objc func getInsights(_ call: CAPPluginCall) {
        guard manager.isAvailable else {
            fail(call, "HealthKit is not available on this device.")
            return
        }
        let requested = call.getString("period") ?? "day"
        let period = ["day", "week", "month", "year"].contains(requested) ? requested : "day"
        Task {
            let range = self.manager.periodRange(period)
            // The flag travels with EVERY payload that carries grantedPermissions, not just
            // syncNow's. Leaving it off here is what kept the Insights screen showing
            // "Permission required" on every metric after the dashboard had started reporting
            // real numbers: same list, same comparison, but without the flag the web layer had
            // no way to know the list could not contain a read permission.
            var data: [String: Any] = [
                "period": period,
                "grantedPermissions": self.grantedList(),
                "readAuthorizationIsUnknowable": true
            ]
            data["steps"]              = await self.orNull { try await self.manager.stepsFor(range) }
            data["activeCalories"]     = await self.orNull { try await self.manager.activeCaloriesFor(range) }
            data["distance"]           = await self.orNull { try await self.manager.distanceFor(range) }
            data["workouts"]           = await self.orNull { try await self.manager.workoutCountFor(range) }
            data["heartRate"]          = await self.orNull { try await self.manager.heartRateFor(range) }
            data["sleep"]              = await self.orNull { try await self.manager.sleepSummary(from: range.0, to: range.1) }
            data["hydration"]          = await self.orNull { try await self.manager.hydrationFor(range) }
            data["nutrition"]          = await self.orNull { try await self.manager.nutritionFor(range) }
            data["weight"]             = await self.orNull { try await self.manager.weightPeriodFor(range) }
            data["respiratoryRate"]    = await self.orNull { try await self.manager.latestRespiratoryRate() }
            data["oxygenSaturation"]   = await self.orNull { try await self.manager.latestOxygenSaturation() }
            data["bloodPressure"]      = await self.orNull { try await self.manager.latestBloodPressure() }
            data["bodyTemperature"]    = await self.orNull { try await self.manager.latestBodyTemperature() }
            data["bodyFat"]            = await self.orNull { try await self.manager.latestBodyFat() }
            data["height"]             = await self.orNull { try await self.manager.latestHeight() }
            data["leanBodyMass"]       = await self.orNull { try await self.manager.latestLeanBodyMass() }
            data["basalMetabolicRate"] = await self.orNull { try await self.manager.latestBasalMetabolicRate() }
            self.ok(call, data)
        }
    }

    // MARK: - Writes

    /// ignytWorkoutId is required, not optional — it becomes the HealthKit sync identifier,
    /// which is what stops a re-exported workout appearing twice in Health.
    @objc func saveWorkout(_ call: CAPPluginCall) {
        // Sent as strings by the JS side because epoch millis exceed what a JS number carries
        // safely through the bridge as an int.
        guard let startMillis = Double(call.getString("startTime") ?? ""),
              let endMillis = Double(call.getString("endTime") ?? ""),
              let ignytWorkoutId = call.getString("ignytWorkoutId"), !ignytWorkoutId.isEmpty else {
            fail(call, "saveWorkout requires startTime, endTime, and ignytWorkoutId.")
            return
        }
        let title = call.getString("title") ?? "Workout"
        let activity = Self.activityType(call.getString("type"))
        Task {
            do {
                let id = try await self.manager.saveWorkout(
                    start: Date(timeIntervalSince1970: startMillis / 1000),
                    end: Date(timeIntervalSince1970: endMillis / 1000),
                    title: title, activity: activity, ignytWorkoutId: ignytWorkoutId)
                self.ok(call, ["recordId": id])
            } catch {
                self.fail(call, "saveWorkout failed: \(error.localizedDescription)")
            }
        }
    }

    @objc func saveWeight(_ call: CAPPluginCall) {
        guard let kg = call.getDouble("weightKg"),
              let ignytBodyLogId = call.getString("ignytBodyLogId"), !ignytBodyLogId.isEmpty else {
            fail(call, "saveWeight requires weightKg and ignytBodyLogId.")
            return
        }
        let millis = Double(call.getString("time") ?? "") ?? Date().timeIntervalSince1970 * 1000
        Task {
            do {
                let id = try await self.manager.saveWeight(
                    kg: kg, at: Date(timeIntervalSince1970: millis / 1000), ignytBodyLogId: ignytBodyLogId)
                self.ok(call, ["recordId": id])
            } catch {
                self.fail(call, "saveWeight failed: \(error.localizedDescription)")
            }
        }
    }

    /// Mirrors mapIgnytExerciseType on Android.
    private static func activityType(_ ignytType: String?) -> HKWorkoutActivityType {
        switch ignytType {
        case "race":   return .highIntensityIntervalTraining
        case "cardio": return .running
        default:       return .traditionalStrengthTraining
        }
    }
}
